### ISSUE TITLE
[QUERY] How to ignore / drop HealthCheck request samples in Application Insights

### ISSUE THREAD

**Issue by dhayanands** [Azure/azure-sdk-for-net#46056]
### Library name and version

Azure.Monitor.OpenTelemetry.Exporter 1.3.0

### Query/Question

Hi Team,

How to drop the samples of health check requests or ignore them when using Azure Monitor Open Telemetry.

using functions .Net 8 Isolated Model and packages Azure.Monitor.OpenTelemetry.Exporter --version 1.3.0 & Microsoft.ApplicationInsights --version 2.22.0

Below are the configurations updated a bit based on [LINK](https://stackoverflow.com/questions/73749565/is-there-a-way-to-filter-out-the-healthcheck-logs-of-an-asp-net-core-6-app-in-az)

```csharp
{
    public class HealthCheck : IHealthCheck
    {
        public static readonly HealthCheck _healthCheck = new HealthCheck();

        public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
        {
            return Task.FromResult(HealthCheckResult.Healthy());
        }

        [Function("HealthCheck")]
        public static async Task<IActionResult> HealthCheckStatus(
                [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "health")] HttpRequest req, FunctionContext context)
        {
            var healthContext = new HealthCheckContext();
            var healthStatus = await _healthCheck.CheckHealthAsync(healthContext);
            return new OkObjectResult(Enum.GetName(typeof(HealthStatus), healthStatus.Status));
        }
    }

    public class CustomTelemetryInitializer : ITelemetryInitializer
    {
        private readonly IHttpContextAccessor _httpContextAccessor;

        public CustomTelemetryInitializer(IHttpContextAccessor httpContextAccessor)
        {
            _httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
        }

        public void Initialize(ITelemetry telemetry)
        {
            var requestPath = _httpContextAccessor.HttpContext?.Request.Path.Value;

            if (string.IsNullOrWhiteSpace(requestPath))
                return;

            // List of paths to sample out
            var pathsToSampleOut = new[] { "/api/health" };

            if (pathsToSampleOut.Any(path => requestPath.Contains(path, StringComparison.OrdinalIgnoreCase)))
            {
                // We don't want to track these requests in the metrics.
                if (telemetry is ISupportAdvancedSampling advancedSampling)
                    advancedSampling.ProactiveSamplingDecision = SamplingDecision.SampledOut;

                // For the case that we cannot filter out the telemetry, we mark it as synthetic
                if (string.IsNullOrWhiteSpace(telemetry.Context.Operation.SyntheticSource))
                    telemetry.Context.Operation.SyntheticSource = "FilteredRequest";
            }
        }
    }

    public class CustomTelemetryProcessor : ITelemetryProcessor
    {
        private readonly ITelemetryProcessor _next;

        public CustomTelemetryProcessor(ITelemetryProcessor next)
        {
            _next = next ?? throw new ArgumentNullException(nameof(next));
        }

        public void Process(ITelemetry item)
        {
            if (!string.IsNullOrEmpty(item.Context.Operation.SyntheticSource)) { return; }
            _next.Process(item);
        }
    }
}
```

have initialized the exporters 

```csharp
var resource = ResourceBuilder.CreateDefault().AddAttributes(resourceAttributes);

TracerProvider tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddSource("WebhookEvent")
    .AddHttpClientInstrumentation()
    .AddAspNetCoreInstrumentation(options =>
    {
        options.Filter = context => !context.Request.Path.Equals("/api/health");
    })
    .SetResourceBuilder(resource)
    .AddAzureMonitorTraceExporter(options =>
    {
        options.ConnectionString = connectionString;
    })
    .Build();

MeterProvider meterProvider = Sdk.CreateMeterProviderBuilder()
    .SetResourceBuilder(resource)
    .AddHttpClientInstrumentation()
    .AddAspNetCoreInstrumentation()
    .AddRuntimeInstrumentation()
    .AddAzureMonitorMetricExporter(options =>
    {
        options.ConnectionString = connectionString;
    })
    .Build();

var host = Host.CreateDefaultBuilder()
    .ConfigureFunctionsWebApplication()
    .ConfigureServices(services =>
    {
        services.AddHttpClient();
        services.AddHealthChecks();
        services.AddSingleton(tracerProvider);
        services.AddSingleton(meterProvider);
        services.AddApplicationInsightsTelemetryWorkerService();
        services.AddSingleton<ITelemetryInitializer, CustomTelemetryInitializer>();
        services.AddSingleton<ITelemetryProcessor, CustomTelemetryProcessor>();
    })
```

![image](https://github.com/user-attachments/assets/6c55a82b-6a01-40df-a841-eb9d38bbfdd4)

### Environment

Azure Functions isolated Model with .Net 8

**Comment by TimothyMothra** [comment]
Hi @dhayanands,
It looks like you're mixing two different telemetry SDKs. This is not recommended.
Please onboard to either one or the other.

Your filtering code looks correct, but I can't begin to troubleshoot your configuration until you switch to using one telemetry SDK or the other.


## Application Insights SDK
This is our classic SDK. This is fully supported and feature rich.

- repo: https://github.com/microsoft/ApplicationInsights-dotnet
- nuget: https://www.nuget.org/packages/Microsoft.ApplicationInsights.aspnetcore
- onboarding guide: https://learn.microsoft.com/azure/azure-monitor/app/asp-net-core

```csharp
services.AddApplicationInsightsTelemetryWorkerService();
services.AddSingleton<ITelemetryInitializer, CustomTelemetryInitializer>();
services.AddSingleton<ITelemetryProcessor, CustomTelemetryProcessor>();
```

## Azure Monitor Exporter & OpenTelemetry
This is our newer SDK which will replace the classic SDK.

- repo: https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter
- nuget: https://www.nuget.org/packages/Azure.Monitor.OpenTelemetry.Exporter
- onboarding: https://learn.microsoft.com/azure/azure-monitor/app/opentelemetry-enable

```csharp
.AddAzureMonitorTraceExporter(options =>
```
```csharp
.AddAzureMonitorMetricExporter(options =>
```

## Functions
I see that you're using Azure Functions.
Please pay attention to their specific onboarding instructions.
- onboarding guide: https://learn.microsoft.com/azure/azure-functions/opentelemetry-howto?tabs=app-insights&pivots=programming-language-csharp#3-enable-opentelemetry-in-your-app

**Comment by gastonmuijtjens** [comment]
We are facing the same problem. We would like to set the synthetic source property from OpenTelemetry. There seems currently no way to set the native `ai.operation.syntheticSource` tag from OpenTelemetry and the Azure Monitor Exporter.

**Comment by dhayanands** [comment]
@TimothyMothra I have updated the code as below.  did not change anything with the CustomTelemetryInitializer & CustomTelemetryProcessor. Still see the same issue. Let me know if im missing anything.

program.cs
```csharp
using Azure.Monitor.OpenTelemetry.AspNetCore;
using Azure.Monitor.OpenTelemetry.Exporter;
using HealthCheck;
using Microsoft.ApplicationInsights.Extensibility;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using OpenTelemetry;

var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddAzureMonitorTraceExporter();

var metricsProvider = Sdk.CreateMeterProviderBuilder()
    .AddAzureMonitorMetricExporter();

var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.AddSimpleConsole(options =>
    {
        options.SingleLine= true;
        //options.IncludeScopes = true;
        //options.TimestampFormat = "HH:mm:ss ";
    });
    builder.AddOpenTelemetry(options =>
    {
        options.AddAzureMonitorLogExporter();
    });
    builder.AddFilter("Microsoft.*", LogLevel.Warning);
    builder.AddFilter("System.Net.Http.HttpClient.*", LogLevel.Warning);
});

var host = Host.CreateDefaultBuilder()
    .ConfigureFunctionsWebApplication()
    .ConfigureServices(services =>
    {
        services.AddHttpClient();
        services.AddHealthChecks();
        services.AddOpenTelemetry().UseAzureMonitor();
        services.AddSingleton(tracerProvider);
        services.AddSingleton(metricsProvider);
        services.AddSingleton(loggerFactory);
        services.AddSingleton<ITelemetryInitializer, CustomTelemetryInitializer>();
        services.AddSingleton<ITelemetryProcessor, CustomTelemetryProcessor>();
    })
    .ConfigureAppConfiguration((context, configuration) =>
    {
        configuration.AddEnvironmentVariables();

        if (context.HostingEnvironment.IsDevelopment())
        {
            configuration.AddJsonFile("local.settings.json", optional: true, reloadOnChange: true);
        }
    })
    .Build();

await host.RunAsync();
```
host.json
```json
{
  "version": "2.0",
  "http": {
    "routePrefix": "api"
  },
  "logging": {
    "applicationInsights": {
      "samplingSettings": {
        "isEnabled": true,
        "excludedTypes": "Request"
      },
      "enableLiveMetricsFilters": true
    }
  },
  "telemetryMode": "openTelemetry"
}
```


![image](https://github.com/user-attachments/assets/cb9ed26b-37d4-461c-af30-aa70aac524de)

![image](https://github.com/user-attachments/assets/84d24ab6-6e26-42ef-8fb1-372d9e96f540)

![image](https://github.com/user-attachments/assets/f2585660-3dfe-4d43-afa7-cd1fc81cf5a4)

**Comment by TimothyMothra** [comment]
> We are facing the same problem. We would like to set the synthetic source property from OpenTelemetry. There seems currently no way to set the native `ai.operation.syntheticSource` tag from OpenTelemetry and the Azure Monitor Exporter.

@gastonmuijtjens, please make a comment on this thread. https://github.com/Azure/azure-sdk-for-net/issues/46021
I'm tracking requests to set specific fields. Our team is working to define syntheticSource in the OpenTelemetry community, so we're waiting to see if that lands.

**Comment by TimothyMothra** [comment]
@dhayanands I was able to reproduce your bug. I'll be working on this with highest priority.

**Comment by dhayanands** [comment]
@johncrim I get traces for 2 different requests: for `http://localhost:50515/api/health` and `https://\<function-name\>.azurewebsites.net/api/health`

when i enable the fileter using `AspNetCoreTraceInstrumentationOptions` it removes the `http://localhost:50515/api/health`  from traces but not the `https://\<function-name\>.azurewebsites.net/api/health`

below is without the filter
![image](https://github.com/user-attachments/assets/259433e2-1074-4381-bd5a-c73a4ae9e6ea)

when i enable the filter it removes the request to localhost but not to the functions url
![image](https://github.com/user-attachments/assets/eebc41c2-ad8b-4f70-b1ad-4d6a497a05c9)

and what i have noticed that the CustomTelemetryInitializer & CustomTelemetryProcessor are not working. I am missing something but not sure what.

**Comment by TimothyMothra** [comment]
@dhayanands what documents are you following to onboard?
I'm seeing a lot of mistakes in your config, if you're following a guide I need to get that corrected.

## Application Insights
`CustomTelemetryInitializer` & `CustomTelemetryProcessor` are both classic Application Insights SDK components and not relevant for OpenTelemetry.

## OpenTelemetry
Your config currently has two OpenTelemetry TracerProviders:

1. `var tracerProvider = Sdk.CreateTracerProviderBuilder()`
     In this provider you've configured a filter configured for incoming requests. 
2. `services.AddOpenTelemetry()`
     This creates an additional TracerProvider which as currently configured does not have a filter.


> when i enable the fileter using AspNetCoreTraceInstrumentationOptions it removes the http://localhost:50515/api/health from traces but not the https://\<function-name\>.azurewebsites.net/api/health

If your filter works for localhost, I would expect it to also work for Functions. 


----
I wrote a brief tutorial so we can get on the same page:

# Tutorial 1: getting familiar with OpenTelemetry

I encourage you to use a minimal app to experiment with the OpenTelemetry SDK so you can get a better understanding of individual components.

The OpenTelemetry repo has several example apps to help you get familiar with the SDK.
- https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/examples

## 1-1 get started with a basic app

Here I've created a basic console app that creates a web application with two endpoints. The console app will query both endpoints and print the response.

```csharp
using Microsoft.AspNetCore.Builder;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.MapGet("/endpoint1", () => "Hello from Endpoint1!");
app.MapGet("/api/health", () => "Hello from /api/health!");

var hostTask = app.RunAsync();

await Task.Delay(1000); // Give the server a moment to start

using var client = new HttpClient();

var response1 = await client.GetStringAsync("http://localhost:5000/endpoint1");
Console.WriteLine(response1);

var response2 = await client.GetStringAsync("http://localhost:5000/api/health");
Console.WriteLine(response2);

await hostTask;
```

OUTPUT:
```
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\tilee\source\repos\ConsoleApp2\ConsoleApp2\bin\Debug\net8.0
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/endpoint1 - - -
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'HTTP: GET /endpoint1'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'HTTP: GET /endpoint1'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://localhost:5000/endpoint1 - 200 - text/plain;+charset=utf-8 109.7035ms
Hello from Endpoint1!
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/api/health - - -
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'HTTP: GET /api/health'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'HTTP: GET /api/health'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://localhost:5000/api/health - 200 - text/plain;+charset=utf-8 9.2923ms
Hello from /api/health!
```


## 1-2 following Azure Functions recommendation, lets add OpenTelemetry

- https://learn.microsoft.com/azure/azure-functions/opentelemetry-howto?tabs=app-insights&pivots=programming-language-csharp#3-enable-opentelemetry-in-your-app

```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOpenTelemetry()
```

This method `AddOpenTelemetry()` will add the `TracerProvider`, `MetricProvider`. The `LoggerFactory` is already included in the `bulider`, but this will configure logging for OpenTelemetry. You were previously doing these three parts manually.

## 1-3 collect incoming requests

By default the OpenTelemetry SDK doesn't collect telemetry. For this you'll need to add the corresponding instrumentation library for the components you want to monitor.

For this that's the AspNetCore instrumentation. And we want to collect requests, so this would be configured with Tracing
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore

```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOpenTelemetry()
    .WithTracing(traces =>
    {
        traces.AddAspNetCoreInstrumentation();
    });
```

## 1-4 output telemetry to the Console

For this simple app, lets output telemetry to the Console using the Console Exporter.
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.Console

```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOpenTelemetry()
    .WithTracing(traces =>
    {
        traces.AddAspNetCoreInstrumentation();
        traces.AddConsoleExporter();
    });
```

Now when we run the console app we see telemetry for those two requests.

OUTPUT
```
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\tilee\source\repos\ConsoleApp2\ConsoleApp2\bin\Debug\net8.0
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/endpoint1 - - -
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'HTTP: GET /endpoint1'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'HTTP: GET /endpoint1'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://localhost:5000/endpoint1 - 200 - text/plain;+charset=utf-8 95.2696ms
Hello from Endpoint1!
Activity.TraceId:            b01d35d8a35ce8f5504471c690c75692
Activity.SpanId:             1e969d812388bab7
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName: Microsoft.AspNetCore
Activity.DisplayName:        GET /endpoint1
Activity.Kind:               Server
Activity.StartTime:          2024-09-20T18:41:09.9680110Z
Activity.Duration:           00:00:00.1008328
Activity.Tags:
    server.address: localhost
    server.port: 5000
    http.request.method: GET
    url.scheme: http
    url.path: /endpoint1
    network.protocol.version: 1.1
    http.route: /endpoint1
    http.response.status_code: 200
Resource associated with Activity:
    telemetry.sdk.name: opentelemetry
    telemetry.sdk.language: dotnet
    telemetry.sdk.version: 1.9.0
    service.name: unknown_service:ConsoleApp2

info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/api/health - - -
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'HTTP: GET /api/health'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'HTTP: GET /api/health'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://localhost:5000/api/health - 200 - text/plain;+charset=utf-8 7.1393ms
Activity.TraceId:            0827b5b69c9928b2347f921f0b4c7408
Activity.SpanId:             36358a9a05b57283
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName: Microsoft.AspNetCore
Activity.DisplayName:        GET /api/health
Activity.Kind:               Server
Activity.StartTime:          2024-09-20T18:41:10.1600518Z
Activity.Duration:           00:00:00.0090729
Activity.Tags:
    server.address: localhost
    server.port: 5000
    http.request.method: GET
    url.scheme: http
    url.path: /api/health
    network.protocol.version: 1.1
    http.route: /api/health
    http.response.status_code: 200
Resource associated with Activity:
    telemetry.sdk.name: opentelemetry
    telemetry.sdk.language: dotnet
    telemetry.sdk.version: 1.9.0
    service.name: unknown_service:ConsoleApp2

Hello from /api/health!
```

## 1-5 filter the specific endpoint.

Now you can add your filter 
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore#filter

```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOpenTelemetry()
    .WithTracing(traces =>
    {
        traces.AddAspNetCoreInstrumentation(options =>
        {
            options.Filter = (httpContext) =>
            {
                return httpContext.Request.Path != "/api/health";
            };
        });
        traces.AddConsoleExporter();
    });
```

Now when you run this app you'll see that requests for the api/health endpoint were correctly removed.

OUTPUT
```
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\tilee\source\repos\ConsoleApp2\ConsoleApp2\bin\Debug\net8.0
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/endpoint1 - - -
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'HTTP: GET /endpoint1'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'HTTP: GET /endpoint1'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://localhost:5000/endpoint1 - 200 - text/plain;+charset=utf-8 101.4578ms
Hello from Endpoint1!
Activity.TraceId:            e5118da2546066346ce2cec66f9b4da9
Activity.SpanId:             fe050b0e15e763c5
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName: Microsoft.AspNetCore
Activity.DisplayName:        GET /endpoint1
Activity.Kind:               Server
Activity.StartTime:          2024-09-20T18:45:22.8734265Z
Activity.Duration:           00:00:00.1095397
Activity.Tags:
    server.address: localhost
    server.port: 5000
    http.request.method: GET
    url.scheme: http
    url.path: /endpoint1
    network.protocol.version: 1.1
    http.route: /endpoint1
    http.response.status_code: 200
Resource associated with Activity:
    telemetry.sdk.name: opentelemetry
    telemetry.sdk.language: dotnet
    telemetry.sdk.version: 1.9.0
    service.name: unknown_service:ConsoleApp2

info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/api/health - - -
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'HTTP: GET /api/health'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'HTTP: GET /api/health'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://localhost:5000/api/health - 200 - text/plain;+charset=utf-8 8.0352ms
Hello from /api/health!
```

# regarding adding Azure Monitor

When you add Azure Monitor, there will be some changes to the code I shared above.

Azure Functions is recommending the `UseAzureMonitor` api. 
This is our distro library that auto-configures a few Instrumentation Libraries that we recommend for web applications.
This means that you won't need to manually call `AddAspNetCoreInstrumentation`.

https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore#what-is-included-in-the-distro

```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddOpenTelemetry().UseAzureMonitor()
```

@johncrim is right, since our package obscures the `AddAspNetCoreInstrumentation` method, you have to use the alternate api to add your filter rule.

- https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore#customizing-aspnetcoretraceinstrumentationoptions
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore#advanced-configuration

```csharp
builder.Services.Configure<AspNetCoreTraceInstrumentationOptions>(options =>
{
    options.Filter = (httpContext) =>
    {
        return httpContext.Request.Path != "/api/health";
    };
});
```

In my sample app, I'm seeing that the filter is being called, but something in the AzureMonitor distro is causing it to be ignored. THIS IS A BUG and I'm currently investigating.

**Comment by TimothyMothra** [comment]
# Follow up.

When I test filtering in a unit test with our Distro project, filtering works.
When I test filtering in a web app with our Distro package, filtering works.
When I test filtering in a console app with our Distro package, filtering does not work.
As we don't support the `Azure.Monitor.OpenTelemetry.AspNetCore` package in a console app, this is not an urgent issue but something we'll pay attention to.

--
@dhayanands, This is the example app that you should use as a reference:

### csproj
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Azure.Monitor.OpenTelemetry.AspNetCore" Version="1.3.0-beta.1" />
    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.9.0" />
  </ItemGroup>

</Project>
```

### Program.cs
```csharp
namespace WebApplication1
{
    using Azure.Monitor.OpenTelemetry.AspNetCore;
    using OpenTelemetry.Instrumentation.AspNetCore;
    using OpenTelemetry.Trace;

    public class Program
    {
        public static void Main(string[] args)
        {
            var builder = WebApplication.CreateBuilder(args);
            builder.Services.AddOpenTelemetry()
                .WithTracing(traces => traces.AddConsoleExporter()) // FOR LOCAL TESTING ONLY
                .UseAzureMonitor(options => options.ConnectionString = "InstrumentationKey=00000");
            
            builder.Services.Configure<AspNetCoreTraceInstrumentationOptions>(options =>
            {
                options.Filter = (httpContext) =>
                {
                    return httpContext.Request.Path != "/api/health";
                };
            });

            var app = builder.Build();

            app.MapGet("/", () => "Hello World!");
            app.MapGet("/api/health", () => "Hello from /api/health!");

            app.Run();
        }
    }
}
```

**Comment by dhayanands** [comment]
@TimothyMothra Thank you. I know my code has something wrong as I were trying to migrate from Application Insights to Open Telemetry including the  filters. But did not know what i missed. I will clean the existing setup and try what you suggested.

**Comment by TimothyMothra** [comment]
@dhayanands I'm sharing a link to our migration guide. This may cover other scenarios you haven't encountered yet: https://learn.microsoft.com/en-us/azure/azure-monitor/app/opentelemetry-dotnet-migrate?tabs=aspnetcore

Of course if you have any other issues, please don't hesitate to engage with us here. :)

**Comment by hansmbakker** [comment]
⚠️Note: when doing
```csharp
builder.Services.Configure<AspNetCoreTraceInstrumentationOptions>(options =>
            {
                options.Filter = (httpContext) =>
                {
                    return httpContext.Request.Path != "/api/health";
                };
            });
```

you might end up seeing parentless errors and traces when an issue happens in the health check or in a dependency. The parent activity (health check API call is not sent, but the exceptions and dependency errors are still sent. This might be confusing - so be aware of that. For this reason, we removed the filtering again.

**Comment by cijothomas** [comment]
> ⚠️Note: when doing
> 
> builder.Services.Configure<AspNetCoreTraceInstrumentationOptions>(options =>
>             {
>                 options.Filter = (httpContext) =>
>                 {
>                     return httpContext.Request.Path != "/api/health";
>                 };
>             });
> you might end up seeing parentless errors and traces when an issue happens in the health check or in a dependency. The parent activity (health check API call is not sent, but the exceptions and dependency errors are still sent. This might be confusing - so be aware of that. For this reason, we removed the filtering again.

Could you open a new issue to track this? There might be alternative ways to do this and at the same time ensure consistent decision for downward http client/other spans.

**Comment by hansmbakker** [comment]
Moved to https://github.com/Azure/azure-sdk-for-net/issues/46751

**Comment by hansmbakker** [comment]
@cijothomas done
